src/SFA.DAS.EmployerIncentives.Data/ApprenticeshipIncentives/ApprenticeshipIncentiveDataRepository.cs

```diff
@@ -151,6 +129,34 @@ public async Task Delete(ApprenticeshipIncentiveModel apprenticeshipIncentive)
             }
         }
 
+        private async Task<ApprenticeshipIncentiveModel> Get(Expression<Func<ApprenticeshipIncentive, bool>> predicate)
+        {
+            var apprenticeshipIncentive = await _dbContext.ApprenticeshipIncentives.FirstOrDefaultAsync(predicate);
+
+            if (apprenticeshipIncentive != null)
+            {
+                await PopulateApprenticeshipIncentive(apprenticeshipIncentive);
+                return apprenticeshipIncentive?.Map(_dbContext.CollectionPeriods.AsEnumerable());
+            }
+
+            return null;
+        }
+
+        private async Task PopulateApprenticeshipIncentive(ApprenticeshipIncentive apprenticeshipIncentive)
+        {
+            apprenticeshipIncentive.PendingPayments = await _dbContext.PendingPayments.Where(x => x.ApprenticeshipIncentiveId == apprenticeshipIncentive.Id).ToListAsync();
+            foreach (var pendingPayment in apprenticeshipIncentive.PendingPayments)
+            {
+                pendingPayment.ValidationResults = await _dbContext.PendingPaymentValidationResults.Where(x => x.PendingPaymentId == pendingPayment.Id).ToListAsync();
+            }
+            apprenticeshipIncentive.Payments = await _dbContext.Payments.Where(x => x.ApprenticeshipIncentiveId == apprenticeshipIncentive.Id).ToListAsync();
+            apprenticeshipIncentive.ClawbackPayments = await _dbContext.ClawbackPayments.Where(x => x.ApprenticeshipIncentiveId == apprenticeshipIncentive.Id).ToListAsync();
+            apprenticeshipIncentive.BreakInLearnings = await _dbContext.BreakInLearnings.Where(x => x.ApprenticeshipIncentiveId == apprenticeshipIncentive.Id).ToListAsync();
+            apprenticeshipIncentive.EmploymentChecks = await _dbContext.EmploymentChecks.Where(x => x.ApprenticeshipIncentiveId == apprenticeshipIncentive.Id).ToListAsync();
+            apprenticeshipIncentive.ValidationOverrides = await _dbContext.ValidationOverrides.Where(x => x.ApprenticeshipIncentiveId == apprenticeshipIncentive.Id).ToListAsync();
```

Comment:
one way to improve this method is to run all queries in parallel but just adding AwaitAll will not be enough you need to create datacontext factory to create a new copy of DBContect otherwise using same instance of context means they will probably run one after the other

Reply:
https://stackoverflow.com/questions/41749896/ef-6-how-to-correctly-perform-parallel-queries

Reply:
https://github.com/SkillsFundingAgency/das-payments-V2-matched-learner/blob/master/src/SFA.DAS.Payments.MatchedLearner.Data/Contexts/MatchedLearnerDataContextFactory.cs

Reply:
See comments about the separate database contexts and EF unit of work

Reply:
Is it not possible to do something like this ?

`
var paymentsTask = _dbContext.Payments.Where(x => x.ApprenticeshipIncentiveId == apprenticeshipIncentive.Id).AsNoTracking().ToListAsync();
            var clawbackPaymentsTask = _dbContext.Payments.Where(x => x.ApprenticeshipIncentiveId == apprenticeshipIncentive.Id).AsNoTracking().ToListAsync();

            await Task.WhenAll(paymentsTask, clawbackPaymentsTask);

            _dbContext.AttachRange(paymentsTask.Result);
            _dbContext.AttachRange(clawbackPaymentsTask.Result);
`

Reply:
Have just tried that - get the following exception when running the first `_dbContext.AttachRange()` :

`System.InvalidOperationException : The instance of entity type 'Payment' cannot be tracked because another instance with the same key value for {'Id'} is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the conflicting key values.`